% Define the top matter
\setModuleTitle{RNA-Seq}
\setModuleAuthors{%
  Myrto Kostadima, EMBL-EBI \mailto{kostadmi@ebi.ac.uk}\\
  Remco Loos, EMBL-EBI \mailto{remco@ebi.ac.uk}
}
\setModuleContributions{%
  Nathan S. Watson-Haigh \mailto{nathan.watson-haigh@awri.com.au} \\
  Sonika Tyagi \mailto{sonika.tyagi@agrf.org.au}%
  Susan M Corley \mailto{susan}%
}

% Start: Module Title Page
\chapter{\moduleTitle}
\newpage
% End: Module Title Page

\section{Key Learning Outcomes}

After completing this practical the trainee should be able to:
\begin{itemize}
  \item Understand and perform a simple RNA-Seq analysis workflow.
  \item Perform gapped alignments to an indexed reference genome using TopHat.
  \item Perform transcript assembly using Cufflinks.
  \item Visualize transcript alignments and annotation in a genome browser such as IGV.
  \item Be able to identify differential gene expression between two experimental conditions.
  \item Be familiar with R environment and be able to run R based RNA-seq packages.
\end{itemize}

\section{Resources You'll be Using}
 
\subsection{Tools Used}
\begin{description}[style=multiline,labelindent=0cm,align=left,leftmargin=1cm]
  \item[Tophat] \hfill\\
    \url{http://tophat.cbcb.umd.edu/}
  \item[Cufflinks] \hfill\\
    \url{http://cufflinks.cbcb.umd.edu/}
  \item[Samtools] \hfill\\
    \url{http://samtools.sourceforge.net/}
  \item[BEDTools] \hfill\\
    \url{http://code.google.com/p/bedtools/}
  \item[UCSC tools] \hfill\\
    \url{http://hgdownload.cse.ucsc.edu/admin/exe/}
  \item[IGV] \hfill\\
    \url{http://www.broadinstitute.org/igv/}
  \item[DAVID Functional Analysis] \hfill\\
    \url{http://david.abcc.ncifcrf.gov/}
  \item[edgeR pakcage] \hfill\\
    \url{http://http://www.bioconductor.org/packages/release/bioc/html/edgeR.html/}
\end{description}

\subsection{Sources of Data}
\url{http://www.ebi.ac.uk/ena/data/view/ERR022484}\\
\url{http://www.ebi.ac.uk/ena/data/view/ERR022485}
\url{http://www.pnas.org/content/suppl/2008/12/16/0807121105.DCSupplemental}

\newpage

\section{Introduction}
The goal of this hands-on session is to perform some basic tasks in the
downstream analysis of RNA-seq data. We will start from RNA-seq data aligned to
the zebrafish genome using Tophat.

We will perform transcriptome reconstruction using Cufflinks and we will compare
the gene expression between two different conditions in order to identify
differentially expressed genes.

In the second part of the tutorial we will also be demonstrating usage of R-based packages to perform differential expression analysis.
We will be using edgeR for the demonstration. The gene/tag counts generated from the alignment are used as input for edgeR.
\section{Prepare the Environment}
We will use a dataset derived from sequencing of mRNA from \textit{Danio rerio} embryos
in two different developmental stages. Sequencing was performed on the Illumina
platform and generated 76bp paired-end sequence data using polyA selected RNA.
Due to the time constraints of the practical we will only use a subset of the
reads.

The data files are contained in the subdirectory called \texttt{data} and are
the following:
\begin{description}[style=multiline,labelindent=1.5cm,align=left,leftmargin=2.5cm]
  \item[\texttt{2cells\_1.fastq} and \texttt{2cells\_2.fastq}] \hfill\\
 These files are based on RNA-seq data of a 2-cell zebrafish embryo
  \item[\texttt{6h\_1.fastq} and \texttt{6h\_2.fastq}] \hfill\\
 These files are based on RNA-seq data of zebrafish embryos 6h post
 fertilization
\end{description}

\begin{steps}
Open the Terminal and go to the \texttt{RNA-seq} working directory:
\begin{lstlisting}
cd ~/RNA-seq/
\end{lstlisting}
\end{steps}

%\reversemarginpar\marginpar{\vskip+0em\hfill\includegraphics[height=1cm]{./graphics/warning.png}}
%\textcolor{red}{
\begin{warning}
  All commands entered into the terminal for this tutorial should be from within the
  \textbf{\texttt{$\sim$/RNA-seq}} directory.
\end{warning}

\begin{steps}
Check that the \texttt{data} directory contains the above-mentioned files by typing:
\begin{lstlisting}
ls data
\end{lstlisting}
\end{steps}

\section{Alignment}
There are numerous tools for performing short read alignment and the choice of aligner
should be carefully made according to the analysis goals/requirements. Here we will
use Tophat, a widely used ultrafast aligner that performs spliced alignments.

Tophat is based on the Bowtie aligner and uses an indexed genome for the
alignment to speed up the alignment and keep its memory footprint small. 
The the index for the \textit{Danio rerio} genome has been created for you. 

\begin{warning}
The command to create an index is as follows. You DO NOT need to run this command
yourself - we have done this for you.
\begin{lstlisting}
bowtie-build genome/Danio_rerio.Zv9.66.dna.fa genome/ZV9
\end{lstlisting}
\end{warning}

\begin{steps}
Tophat has a number of parameters in order to perform the alignment. To view them all type:
\begin{lstlisting}
tophat --help
\end{lstlisting}
\end{steps}

\begin{information}
The general format of the tophat command is:
\begin{lstlisting}[style=command_syntax]
tophat [options]* <index_base> <reads_1> <reads_2>
\end{lstlisting}

Where the last two arguments are the \texttt{.fastq} files of the paired end
reads, and the argument before is the basename of the indexed genome.
\end{information}

\begin{note}
The quality values in the FASTQ files used in this hands-on session are Phred+33
encoded. We explicitly tell tophat of this fact by using the command line
argument \texttt{--solexa-quals}.
\end{note}

\begin{information}
You can look at the first few reads in the file \texttt{data/2cells\_1.fastq} with:
 
\begin{lstlisting}
head -n 20 data/2cells_1.fastq
\end{lstlisting}
\end{information}

%\reversemarginpar\marginpar{\vskip+0em\hfill\includegraphics[height=1cm]{./graphics/notes.png}}
\begin{note}
Some other parameters that we are going to use to run Tophat are listed below:
\begin{description}[style=multiline,labelindent=0cm,align=right,leftmargin=\descriptionlabelspace,rightmargin=1.5cm,font=\ttfamily]
 \item[-g] Maximum number of multihits allowed. Short reads are likely to map to
 more than one location in the genome even though these reads can have originated
 from only one of these regions. In RNA-seq we allow for a limited number of
 multihits, and in this case we ask Tophat to report only reads that map at most
 onto 2 different loci.
 \item[--library-type] Before performing any type of RNA-seq analysis you need
 to know a few things about the library preparation. Was it done using a
 strand-specific protocol or not? If yes, which strand? In our data the protocol
 was NOT strand specific.
 \item[-j] Improve spliced alignment by providing Tophat with annotated splice
 junctions. Pre-existing genome annotation is an advantage when analysing RNA-seq
 data. This file contains the coordinates of annotated splice junctions from Ensembl.
 These are stored under the sub-directory \texttt{annotation} in a file called
 \texttt{ZV9.spliceSites}.
 \item[-o] This specifies in which subdirectory Tophat should save the output
 files. Given that for every run the name of the output files is the same, we
 specify different directories for each run.
\end{description}
\end{note}

It takes some time (approx. 20 min) to perform tophat spliced alignments, even for this subset of
reads. Therefore, we have pre-aligned the \texttt{2cells} data for you using the following command:
\begin{warning}
You DO NOT need to run this command yourself - we have done this for you.

\begin{lstlisting}
tophat --solexa-quals -g 2 --library-type fr-unstranded -j annotation/Danio_rerio.Zv9.66.spliceSites -o tophat/ZV9_2cells genome/ZV9 data/2cells_1.fastq data/2cells_2.fastq
\end{lstlisting}
\end{warning}

\begin{steps}
Align the \texttt{6h} data yourself using the following command:  

\begin{lstlisting}
# Takes approx. 20mins
tophat --solexa-quals -g 2 --library-type fr-unstranded -j annotation/Danio_rerio.Zv9.66.spliceSites -o tophat/ZV9_6h genome/ZV9 data/6h_1.fastq data/6h_2.fastq
\end{lstlisting}

\end{steps}

The \texttt{6h} read alignment will take approx. 20 min to complete. Therefore,
we'll take a look at some of the files, generated by tophat, for the
pre-computed \texttt{2cells} data.

\subsection{Alignment Visualisation in IGV}

The Integrative Genomics Viewer (IGV) is able to provide a visualisation of read
alignments given a reference sequence and a BAM file. We'll visualise the
information contained in the \texttt{accepted\_hits.bam} and
\texttt{junctions.bed} files for the pre-computed \texttt{2cells} data. The
former, contains the tophat sliced alignments of the reads to the reference
while the latter stores the coordinates of the splice junctions present in the
data set.

\begin{steps}
Open the \texttt{RNA-seq} directory on your Desktop and double-click the
\texttt{tophat} subdirectory and then the \texttt{ZV9\_2cells} directory.

\begin{enumerate}
  \item Launch IGV by double-clicking the ``IGV 2.3.*'' icon on the Desktop
  (ignore any warnings that you may get as it opens). \emph{NOTE: IGV may take
  several minutes to load for the first time, please be patient.}
  \item Choose ``Zebrafish (Zv9)'' from the drop-down box in the top left of the
  IGV window. Else you can also load the genome fasta file.
  \item Load the \texttt{accepted\_hits.sorted.bam} file by clicking the
  ``File'' menu, selecting ``Load from File'' and navigating to the
  \texttt{Desktop/RNA-seq/tophat/ZV9\_2cells} directory.
  \item Rename the track by right-clicking on its name and choosing ``Rename
  Track''. Give it a meaningful name like ``2cells BAM''.
  \item Load the \texttt{junctions.bed} from the same directory and rename the
  track ``2cells Junctions BED''.
  \item Load the Ensembl annotations file \texttt{Danio\_rerio.Zv9.66.gtf}
  stored in the \texttt{RNA-seq/annotation} directory.
  \item Navigate to a region on chromosome 12 by typing
  \texttt{chr12:20,270,921-20,300,943} into the search box at the top of the IGV
  window.
\end{enumerate}

\end{steps}

\begin{information}
Keep zooming to view the bam file alignments

Some useful IGV manuals can be found below

\url{http://www.broadinstitute.org/software/igv/interpreting_insert_size}\\
\url{http://www.broadinstitute.org/software/igv/alignmentdata}
\end{information}


\begin{questions}
Can you identify the splice junctions from the BAM file?
\begin{answer}
Slice junctions can be identified in the alignment BAM files.
These are the aligned RNA-Seq reads that have skipped-bases from the reference genome (most likely introns).
\end{answer}

Are the junctions annotated for \texttt{CBY1} consistent with the annotation?
\begin{answer}
Read alignment supports an extended length in exon 5 to the gene model (cby1-001) 
\end{answer}

Are all annotated genes, from both RefSeq and Ensembl, expressed?
\begin{answer}
No BX000473.1-201 is not expressed
\end{answer}

\end{questions}

\begin{steps}
Once tophat finishes aligning the 6h data you will need to sort the alignments found in the BAM file and then index the
sorted BAM file.

\begin{lstlisting}
samtools sort tophat/ZV9_6h/accepted_hits.bam tophat/ZV9_6h/accepted_hits.sorted
samtools index tophat/ZV9_6h/accepted_hits.sorted.bam
\end{lstlisting}

Load the sorted BAM file into IGV, as described previously, and rename the track appropriately.
\end{steps}

\newpage
\section{Isoform Expression and Transcriptome Assembly}
There are a number of tools that perform reconstruction of the transcriptome
and for this workshop we are going to use Cufflinks. Cufflinks can do
transcriptome assembly either \textit{ab initio} or using a reference annotation. It
also quantifies the isoform expression in Fragments
Per Kilobase of exon per Million fragments mapped (FPKM).

\begin{steps}
Cufflinks has a number of parameters in order to perform transcriptome
assembly and quantification. To view them all type:

\begin{lstlisting}
cufflinks --help
\end{lstlisting}
\end{steps}

We aim to reconstruct the transcriptome for both samples by using the Ensembl
annotation both strictly and as a guide. In the first case Cufflinks will only
report isoforms that are included in the annotation, while in the latter case
it will report novel isoforms as well.

The Ensembl annotation for \textit{Danio rerio} is available in
\texttt{annotation/Danio\_rerio.Zv9.66.gtf}.

\begin{information}
The general format of the \texttt{cufflinks} command is:
\begin{lstlisting}[style=command_syntax]
cufflinks [options]* <aligned_reads.(sam|bam)>
\end{lstlisting}
Where the input is the aligned reads (either in SAM or BAM format).
\end{information}

\begin{note}
Some of the available parameters for Cufflinks that we are going to use to run
Cufflinks are listed below:
\begin{description}[style=multiline,labelindent=0cm,align=right,leftmargin=\descriptionlabelspace,rightmargin=1.5cm,font=\ttfamily]
  \item[-o] Output directory.
  \item[-G] Tells Cufflinks to use the supplied GTF annotations strictly in order
  to estimate isoform annotation.
  \item[-b] Instructs Cufflinks to run a bias detection and correction algorithm
  which can significantly improve accuracy of transcript abundance estimates.
  To do this Cufflinks requires a multi-fasta file with the genomic sequences
  against which we have aligned the reads.
  \item[-u] Tells Cufflinks to do an initial estimation procedure to more
  accurately weight reads mapping to multiple locations in the genome
  (multi-hits).
  \item[--library-type] Before performing any type of RNA-seq analysis you need
  to know a few things about the library preparation. Was it done using a
  strand-specific protocol or not? If yes, which strand? In our data the protocol
  was NOT strand specific.
\end{description}
\end{note}

\begin{steps}
Perform transcriptome assembly, strictly using the supplied GTF annotations, for the \texttt{2cells} and \texttt{6h} data using cufflinks:
\begin{lstlisting}
# 2cells data (takes approx. 5mins):
cufflinks -o cufflinks/ZV9_2cells_gtf -G annotation/Danio_rerio.Zv9.66.gtf -b genome/Danio_rerio.Zv9.66.dna.fa -u --library-type fr-unstranded tophat/ZV9_2cells/accepted_hits.bam
# 6h data (takes approx. 5mins):
cufflinks -o cufflinks/ZV9_6h_gtf -G annotation/Danio_rerio.Zv9.66.gtf -b genome/Danio_rerio.Zv9.66.dna.fa -u --library-type fr-unstranded tophat/ZV9_6h/accepted_hits.bam
\end{lstlisting}
\end{steps}

\begin{information}
Cufflinks generates several files in the specified output directory. Here's a short description of these files:

\begin{description}[style=multiline,labelindent=0cm,align=right,leftmargin=\descriptionlabelspace,rightmargin=1.5cm,font=\ttfamily]
\item[genes.fpkm\_tracking] Contains the estimated gene-level expression values.
\item[isoforms.fpkm\_tracking] Contains the estimated isoform-level expression values.
\item[skipped.gtf] Contains loci skipped as a result of exceeding the maximum number of fragments.
\item[transcripts.gtf] This GTF file contains Cufflinks' assembled isoforms.
\end{description}

The complete documentation can be found at:
\url{http://cufflinks.cbcb.umd.edu/manual.html#cufflinks_output}
\end{information}

\begin{information}
So far we have forced cufflinks, by using the \texttt{-G} option, to strictly
use the GTF annotations provided and thus novel transcripts will not be reported. We
can get cufflinks to perform a GTF-guided transcriptome assembly by using the
\texttt{-g} option instead. Thus, novel transcripts will be reported.

\end{information}

\begin{warning}
GTF-guided transcriptome assembly is more computationally intensive than
strictly using the GTF annotations. Therefore, we have pre-computed these
GTF-guided assemblies for you and have placed the results under subdirectories:
\texttt{cufflinks/ZV9\_2cells\_gtf\_guided} and
\texttt{cufflinks/ZV9\_6h\_gft\_guided}.

You DO NOT need to run these commands. We provide them so you know how we
generated the the GTF-guided transcriptome assemblies:
\begin{lstlisting}
# 2cells guided transcriptome assembly (takes approx. 30mins):
cufflinks -o cufflinks/ZV9_2cells_gtf_guided -g annotation/Danio_rerio.Zv9.66.gtf -b genome/Danio_rerio.Zv9.66.dna.fa -u --library-type fr-unstranded tophat/ZV9_2cells/accepted_hits.bam
# 6h guided transcriptome assembly (takes approx. 30mins):
cufflinks -o cufflinks/ZV9_6h_gtf_guided -g annotation/Danio_rerio.Zv9.66.gtf -b genome/Danio_rerio.Zv9.66.dna.fa -u --library-type fr-unstranded tophat/ZV9_6h/accepted_hits.bam
\end{lstlisting}

\end{warning}

\begin{steps}
\begin{enumerate}
  \item Go back to IGV and load the pre-computed, GTF-guided transcriptome
  assembly for the \texttt{2cells} data
  (\texttt{cufflinks/ZV9\_2cells\_gtf\_guided/transcripts.gtf}).
  \item Rename the track as ``2cells GTF-Guided Transcripts''.
  \item In the search box type \texttt{ENSDART00000082297} in order for the
  browser to zoom in to the gene of interest.
\end{enumerate}
\end{steps}

\begin{questions}
Do you observe any difference between the Ensembl GTF annotations and the
GTF-guided transcripts assembled by cufflinks (the ``2cells GTF-Guided Transcripts'' track)?
\begin{answer}
Yes. It appears that the Ensembl annotations may have truncated the last exon.
However, our data also doesn't contain reads that span between the last two
exons.
\end{answer}

\end{questions}

\section{Differential Expression}
One of the stand-alone tools that perform differential expression analysis is
Cuffdiff. We use this tool to compare between two conditions; for example
different conditions could be control and disease, or wild-type and mutant, or
various developmental stages.

In our case we want to identify genes that are
differentially expressed between two developmental stages; a \texttt{2cells}
embryo and \texttt{6h} post fertilization.

\begin{information}
The general format of the cuffdiff command is:
\begin{lstlisting}[style=command_syntax]
cuffdiff [options]* <transcripts.gtf> <sample1_replicate1.sam[,...,sample1_replicateM]> <sample2_replicate1.sam[,...,sample2_replicateM.sam]>
\end{lstlisting}

Where the input includes a \texttt{transcripts.gtf} file, which is an annotation
file of the genome of interest or the cufflinks assembled transcripts, and the aligned reads (either in SAM or BAM
format) for the conditions.
Some of the Cufflinks options that we will use to run the program are:
\begin{description}[style=multiline,labelindent=0cm,align=right,leftmargin=\descriptionlabelspace,rightmargin=1.5cm,font=\ttfamily]
  \item[-o] Output directory.
  \item[-L] Labels for the different conditions
  \item[-T] Tells Cuffdiff that the reads are from a time series experiment.
  \item[-b] Instructs Cufflinks to run a bias detection and correction algorithm
  which can significantly improve accuracy of transcript abundance estimates.
  To do this Cufflinks requires a multi-fasta file with the genomic sequences
  against which we have aligned the reads.
  \item[-u] Tells Cufflinks to do an initial estimation procedure to more
  accurately weight reads mapping to multiple locations in the genome
  (multi-hits). 
  \item[--library-type] Before performing any type of RNA-seq analysis you need
 to know a few things about the library preparation. Was it done using a
 strand-specific protocol or not? If yes, which strand? In our data the protocol
 was NOT strand specific.
  \item[-C] biological replicates and multiple group contrast can be defined here 
\end{description}
\end{information}

\begin{steps}
Run cuffdiff on the tophat generated BAM files for the 2cells vs. 6h data sets:
\begin{lstlisting}
cuffdiff -o cuffdiff/ -L ZV9_2cells,ZV9_6h -T -b genome/Danio_rerio.Zv9.66.dna.fa -u --library-type fr-unstranded annotation/Danio_rerio.Zv9.66.gtf tophat/ZV9_2cells/accepted_hits.bam tophat/ZV9_6h/accepted_hits.bam
\end{lstlisting}
\end{steps}

\begin{steps}
We are interested in the differential expression at the gene level. The results
are reported by Cuffdiff in the file \texttt{cuffdiff/gene\_exp.diff}. 
Look at the first few lines of the file using the following command:
\begin{lstlisting}
head -n 20 cuffdiff/gene_exp.diff
\end{lstlisting}

We would like to see which are the most significantly differentially expressed
genes. Therefore we will sort the above file according to the q value
(corrected p value for multiple testing). The result will be stored in a
different file called \texttt{gene\_exp\_qval.sorted.diff}.
\begin{lstlisting}
sort -t$'\t' -g -k 13 cuffdiff/gene_exp.diff > cuffdiff/gene_exp_qval.sorted.diff
\end{lstlisting}

Look again at the first few lines of the sorted file by typing:
\begin{lstlisting}
head -n 20 cuffdiff/gene_exp_qval.sorted.diff
\end{lstlisting}

Copy an Ensembl transcript identifier from the first two columns for one of
these genes (e.g. \texttt{ENSDARG00000077178}). Now go back to the IGV browser
and paste it in the search box.
\end{steps}

\begin{questions}
What are the various outputs generated by cuffdiff ? 
Hint:Please refer to the  \texttt{Cuffdiff output} section of the cufflinks manual online.

Do you see any difference in the read coverage between the \texttt{2cells} and
\texttt{6h} conditions that might have given rise to this transcript being
called as differentially expressed?
\begin{warning}
Note that the coverage on the Ensembl browser is based on raw reads and no
normalisation has taken place contrary to the FPKM values.
\end{warning}

\begin{answer}
The read coverage of this transcript (\texttt{ENSDARG00000077178}) in the 2cells
data set is much higher than in the 6h data set.
\end{answer}

\end{questions}

\begin{information}
\texttt{Cuffquant} utility from te cufflinks suite can be used to generate the count files to be used with count based differential analysis methods such as, \texttt{edgeR} and \texttt{Deseq}.
\end{information}

\section{Functional Annotation of Differentially Expressed Genes}
After you have performed the differential expression analysis you are interested
in identifying if there is any functionality enrichment for your differentially
expressed genes.
On your Desktop click:
\begin{lstlisting}[style=command_syntax]
Applications >> Internet >> Firefox Web Browser
\end{lstlisting}
And go to the following URL: \url{http://david.abcc.ncifcrf.gov/}
On the left side click on Functional Annotation. Then click on the Upload tab.
Under the section Choose from File, click Choose File and navigate to the
\texttt{cuffdiff} directory. Select the file called \texttt{globalDiffExprs\_Genes\_qval.01\_top100.tab}.
Under Step 2 select ENSEMBL\_GENE\_ID from the drop-down menu. Finally select
Gene list and then press Submit List.
Click on Gene Ontology and then click on the CHART button of the GOTERM\_BP\_ALL item.

\begin{questions}
Do these categories make sense given the samples we're studying?
\begin{answer}
Developmental Biology
\end{answer}

Browse around DAVID website and check what other information are available.
\begin{answer}
Cellular component, Molecular function, Biological Processes, Tissue expression, Pathways, Literature, Protein domains 
\end{answer}
\end{questions}


\section{Differential Gene Expression Analysis using edgeR}
\begin{information}
The example we are working through today follows a case Study set out in the edgeR Users Guide (4.3 Androgen-treated prostate cancer cells (RNA-Seq, two groups) which is based on an experiment conducted by Li et al. (2008, Proc Natl Acad Sci USA, 105, 20179-84).
The researches used a prostate cancer cell line (LNCaP cells). These cells are sensitive to stimulation by male hormones (androgens). 
Three replicate RNA samples were collected from LNCaP cells treated with an androgen hormone (DHT). Four replicates were collected from cells treated with an inactive compound. 
Each of the seven samples was run on a lane (7 lanes) of an Illumina flow cell to produce 35 bp reads.
The experimental design was therefore:

\begin{table}[H]
  \centering
  \caption{Experimental design}
    \begin{tabular}{rrr}
    \toprule
    \textbf{Lane)} & \textbf{Treatment} & \textbf{Label} \\
    \midrule
    1    & Control     & Con1\% \\
    2    & Control    & Con2\% \\
    3    & Control & Con3\% \\
    4    & Control & Con4\% \\
    5    & DHT & DHT1\% \\
    5    & DHT & DHT2\% \\
    5    & DHT & DHT3\% \\

    \bottomrule
    \end{tabular}
  \label{tab:experimental design}
\end{table}



\end{information}
\begin{steps}
\begin{enumerate}
  \item Load libraries
\begin{lstlisting}
library(edgeR)
library(biomaRt)
library(gplots)
\end{lstlisting}

\item read in count table and experimental design
\begin{lstlisting}
data <- read.delim("pnas_expression.txt", row.names=1, header=T)
targets <- read.delim("Targets.txt", header=T)
\end{lstlisting}
\item Create DGEList object
\begin{lstlisting}
Y <- DGEList(counts=data[,1:7], group=targets$Treatment)
\end{lstlisting}
\item Change the column names of the object to align with treatment
\begin{lstlisting}
colnames(y) <- targets$Label
\end{lstlisting}
\item Check the dimensions of the object
\begin{lstlisting}
dim(y)
\end{lstlisting}

\item We see we have 37435 rows (i.e. genes) and 7 columns (samples)
Now we will filter out genes with low counts by only keeping those rows where the count per million (cpm) is at least 1 in at least three samples

\begin{lstlisting}
keep <-rowSums( cpm(y)>1) >=3
y <- y[keep, ]

\end{lstlisting}

\end{steps}


\begin{questions}
How many rows (genes) are retained now
\begin{answer} dim(y) would give you 16494

How many genes were filtered out?
\begin{answer} 37435-16494
\end{questions}

\begin{steps}
\item We will now perform normalization to take account of different library size

\begin{lstlisting}
y<-calcNormFactors(y)
\end{lstlisting}

\item We will check the calculated normalization factors
\begin{lstlisting}
y$samples
\end{lstlisting}


\item Lets have a look at whether the samples cluster by condition. (You should produce a plot as shown in Figure )
\begin{lstlisting}
plotMDS(y)
\end{lstlisting}

\begin{figure}[H]
\centering
\includegraphics[width=0.8\textwidth]{rna-seq/mds.png}
\caption{Visualization of smaple clustering}
\label{fig:MDS plot}
\end{figure}

\end{steps}

\begin{information}
We see here that the common dispersion estimates the overall Biological Coefficient of Variation (BCV) of the dataset averaged over all genes. The common dispersion is \texttt{0.02} and the BCV is the square root of the common dispersion \texttt{(/sqrt0.02 = 0.14)}.  A BCV of 14\% is typical for cell line experiment.

\end{information}

\begin{steps}
\item We now estimate gene-specific dispersion

\begin{lstlisting}
y <- estimateTagwiseDisp(y) 
\end{lstlisting}
\item We will plot the tagwise dispersion and the common dispersion (You should obtain a plot as shown in Figure )
\begin{lstlisting}
plotBCV(y)
\end{lstlisting}

\end{steps}
\begin{figure}[H]
\centering
\includegraphics[width=0.8\textwidth]{rna-seq/bcv.png}
\caption{Visualization of smaple clustering}
\label{fig:MDS plot}
\end{figure}

\begin{steps}
\item We now test for differentially expressbcvgenes
\begin{lstlisting}
et <- exactTest(y)
\end{lstlisting}


\item Now we will use the topTags function to adjust for multiple testing. We will use the Benjimini Hochberg (“BH”) method and we will produce a table of results
\begin{lstlisting}
res <- topTags(et, n=nrow(y$counts), adjust.method=”BH”)$table
\end{lstlisting}

\item Let's have a look at the first rows of the table

\begin{lstlisting}
head(res)
\end{lstlisting}

\end{steps}
\begin{information}
You can see we have the ensemble gene identifier in the first column, the log fold change in the second column, the the logCP, the PValue and the adjusted PValue. 
The ensemble gene identifier is not as helpful as the gene symbol so let’s add in a column with the gene symbol. We will use the BiomaRt package to do this.
\end{information}


\begin{steps}
\item We start by using the useMart function of BiomaRt to access the human data base of ensemble gene ids.  Then we create a vector of our ensemble gene ids.
\begin{lstlisting}
ensembl=useMart("ensembl", dataset="hsapiens_gene_ensembl")
ensembl_names<-rownames(res)
\end{lstlisting}

\item We then use the function getBM to get the gene symbol data we want

\begin{lstlisting}
genemap <-getBM( attributes= c("ensembl_gene_id", "entrezgene", "hgnc_symbol"), filters="ensembl_gene_id", values=ensembl_names, mart=ensembl)
\end{lstlisting}

This can taken about a minute or so.

\item Have a look at the start of the genemap dataframe.

\begin{lstlisting}
head(genemap)
\end{lstlisting}


\item We see that we have 3 columns, the ensemble id, the entrez gene id and the hgnc symbol
We use the match function to match up our data with the data we have just retrieved from the database.

\begin{lstlisting}
idx <- match(ensembl_names, genemap$ensembl_gene_id )
res$entrez <-genemap$entrezgene [ idx ]
res$hgnc_symbol <- genemap$hgnc_symbol [ idx ]
\end{lstlisting}


\item Next we have a look at the head of our \texttt{res} dataframe
\begin{lstlisting}
head(res)
\end{lstlisting}
As you see we have now added the hgnc symbol and the entrez id to our results.
\item Let’s now make a subset of the most significant upregulated and downregulated genes.

\begin{lstlisting}
de<-res[res$FDR<0.05, ]
de_upreg <-res[res$FDR<0.05 & res$logFC >0,] 
de_downreg <-res[res$FDR<0.05 & res$logFC <0,] 

\end{lstlisting}
\end{steps}

\begin{questions}
How many differentially expressed genes are there?
\begin{warning} Try str(de)
\end{warning}
\begin{answer} 4429

How many upregulated genes and downregulated genes do we have?
\begin{answer}
str(de_upreg) = 2345
str(de_downreg) = 2084

\end{questions}

\begin{steps}
\item Lets write out these results
\begin{lstlisting}
write.csv( as.data.frame(de), file=”DE_results.csv”)
\end{lstlisting}

\end{enumerate}
\end{steps}


\begin{bonus}
\end{bonus}

\newpage

\section{References}
%TODO Change to using BiBTeX
\begin{enumerate}
  \item Trapnell, C., Pachter, L. \& Salzberg, S. L. TopHat: discovering splice
  junctions with RNA-Seq. Bioinformatics 25, 1105-1111 (2009).
  \item Trapnell, C. et al. Transcript assembly and quantification by RNA-Seq
  reveals unannotated transcripts and isoform switching during cell
  differentiation. Nat. Biotechnol. 28, 511-515 (2010).
  \item Langmead, B., Trapnell, C., Pop, M. \& Salzberg, S. L. Ultrafast and
  memory-efficient alignment of short DNA sequences to the human genome.
  Genome Biol. 10, R25 (2009).
  \item Roberts, A., Pimentel, H., Trapnell, C. \& Pachter, L. Identification
  of novel transcripts in annotated genomes using RNA-Seq. Bioinformatics 27,
  2325-2329 (2011).
  \item Roberts, A., Trapnell, C., Donaghey, J., Rinn, J. L. \& Pachter, L.
  Improving RNA-Seq expression estimates by correcting for fragment bias.
  Genome Biol. 12, R22 (2011).
\end{enumerate}
